core/controllers/story_editor.py

```diff
@@ -0,0 +1,165 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for the story editor."""
+from core.controllers import base
+from core.domain import acl_decorators
+from core.domain import story_domain
+from core.domain import story_services
+from core.domain import topic_domain
+from core.domain import topic_services
+import feconf
+import utils
+
+
+def _require_valid_id(entity_id):
+    """Checks whether the id received from the frontend is a valid one.
+    """
+    if not isinstance(entity_id, basestring):
+        raise base.BaseHandler.InvalidInputException(
+            Exception('All ids should be strings.'))
+
+    if len(entity_id) != 12:
+        raise base.BaseHandler.InvalidInputException(
+            Exception('The id given is invalid.'))
+
+
+class StoryEditorPage(base.BaseHandler):
+    """The editor page for a single story."""
+
+    @acl_decorators.can_edit_story
+    def get(self, topic_id, story_id):
+        """Handles GET requests."""
+
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            raise self.PageNotFoundException()
+
+        _require_valid_id(story_id)
+        _require_valid_id(topic_id)
+
+        story = story_services.get_story_by_id(story_id, strict=False)
+
+        if story is None:
+            raise self.PageNotFoundException(
+                Exception('The story with the given id doesn\'t exist.'))
+
+        self.values.update({
+            'story_id': story.id
+        })
+
+        self.render_template('pages/story_editor/story_editor.html')
+
+
+class EditableStoryDataHandler(base.BaseHandler):
+    """A data handler for stories which support writing."""
+
+    def _require_valid_version(self, version_from_payload, story_version):
+        """Check that the payload version matches the given story
+        version.
+        """
+        if version_from_payload is None:
+            raise base.BaseHandler.InvalidInputException(
+                'Invalid POST request: a version must be specified.')
+
+        if version_from_payload != story_version:
+            raise base.BaseHandler.InvalidInputException(
+                'Trying to update version %s of story from version %s, '
+                'which is too old. Please reload the page and try again.'
+                % (story_version, version_from_payload))
+
+    @acl_decorators.can_edit_story
+    def get(self, topic_id, story_id):
+        """Populates the data on the individual story page."""
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            raise self.PageNotFoundException()
+
+        _require_valid_id(story_id)
+        _require_valid_id(topic_id)
+
+        story = story_services.get_story_by_id(story_id, strict=False)
```

Comment:
Ditto here and below for verifying both topic and story. Slows things down a bit but let's play it safe for now and relax the requirement later if it becomes a problem.

Reply:
Done.